hoomd/mpcd/CollisionMethod.cu

```diff
@@ -0,0 +1,263 @@
+    d_initial_vel[idx] = d_velocity[particle_idx];
+    }
+
+__global__ void accumulate_rigid_body_momenta(Scalar3* d_linmom_accum,
+                                              Scalar3* d_angmom_accum,
+                                              const Scalar4* d_initial_vel,
+                                              const unsigned int* d_embed_group,
+                                              const Scalar4* d_postype,
+                                              const Scalar4* d_velocity,
+                                              const int3* d_image,
+                                              const unsigned int* d_body,
+                                              const unsigned int* d_rtag,
+                                              const BoxDim global_box,
+                                              const unsigned int num_group)
+    {
+    // one thread per particle
+    unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx >= num_group)
+        return;
+
+    // get the index from the embedded group and check if in a rigid body
+    unsigned int particle_idx = d_embed_group[idx];
+    const unsigned int central_tag = d_body[particle_idx];
+    if (central_tag >= MIN_FLOPPY)
+        {
+        return;
+        }
+    const unsigned int central_idx = d_rtag[central_tag];
+    // if the central particle is not local, cannot read or write to it.
+    assert(central_idx != NOT_LOCAL);
+
+    // collision on central particle itself already taken care of by collision rule
+    if (particle_idx == central_idx)
+        {
+        return;
+        }
+    // get velocities and masses
+    const Scalar4 vel_mass_const = d_velocity[particle_idx];
+    const vec3<Scalar> vel_const(vel_mass_const);
+    const Scalar mass_const = vel_mass_const.w;
+
+    // get displacement
+    const Scalar4 postype_const = d_postype[particle_idx];
+    const vec3<Scalar> pos_const(postype_const);
+    const int3 img_const = d_image[particle_idx];
+
+    const Scalar4 postype_central = d_postype[central_idx];
+    const vec3<Scalar> pos_central(postype_central);
+    const int3 img_central = d_image[central_idx];
+
+    vec3<Scalar> displacement = pos_const - pos_central;
+    const int3 displacement_img = make_int3(img_const.x - img_central.x,
+                                            img_const.y - img_central.y,
+                                            img_const.z - img_central.z);
+    displacement = global_box.shift(displacement, displacement_img);
+
+    // change in linear and angular momentum
+    const vec3<Scalar> initial_vel_const(d_initial_vel[idx]);
+    const vec3<Scalar> linmom_change = mass_const * (vel_const - initial_vel_const);
+    const vec3<Scalar> angmom_change = cross(displacement, linmom_change);
+
+    // accumulate onto central particle
+    d_linmom_accum[central_idx] += vec_to_scalar3(linmom_change);
+    d_angmom_accum[central_idx] += vec_to_scalar3(angmom_change);
```

Comment:
To get this working, we should be able to replace each `+=` by 3 atomic transactions to avoid the race condition summing onto the particle.
```suggestion
    atomicAdd(&d_linmom_accum[central_idx].x, linmom_change.x);
    atomicAdd(&d_linmom_accum[central_idx].y, linmom_change.y);
    atomicAdd(&d_linmom_accum[central_idx].z, linmom_change.z);
    atomicAdd(&d_angmom_accum[central_idx].x, angmom_change.x);
    atomicAdd(&d_angmom_accum[central_idx].y, angmom_change.y);
    atomicAdd(&d_angmom_accum[central_idx].z, angmom_change.z);
```